### PR TITLE
Add Bonsai C4 converter

### PR DESCRIPTION
**Pull request by infinityis**
Adds converter for:

promicro -> [bonsai c4](https://github.com/customMK/Bonsai-C/tree/main/C4)

At the moment this PR only performs basic pin conversion. Despite peripheral support being needed to access on-board EEPROM and flash memory chips on Bonsai C4, peripheral support is currently not included because this converter leverages the existing GENERIC_STM32_F411XE board definition, and boards inheriting boards is not yet available.

If it is preferable to use dedicated Bonsai C4 board/config files (which include peripheral support) in the meantime, I have that ready as well and can update the PR accordingly.

## Types of Changes

- [x] Core
- [ ] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Review by zvecr — DISMISSED** [review]

**Comment by zvecr** [comment]
Could you also update the docs at <https://github.com/qmk/qmk_firmware/blob/develop/docs/feature_converters.md>.

> If it is preferable to use dedicated Bonsai C4 board/config files (which include peripheral support) in the meantime, I have that ready as well and can update the PR accordingly.

Maybe, not sure I will have time before the end of the current develop cycle to improve the situation. Somewhat depends the level of "Out the box" configuration its providing.

File: docs/feature_converters.md [hunk 1]
```diff
@@ -13,6 +13,7 @@ Currently the following converters are available:
 | `promicro` | `proton_c`        |
 | `promicro` | `kb2040`          |
 | `promicro` | `promicro_rp2040` |
+| `promicro` | `bonasai_c4`      |
```
Comment (tzarc) [inline docs/feature_converters.md:16]:
```suggestion
| `promicro` | `bonsai_c4`      |
```

File: keyboards/custommk/bonsai_c4_template/bonsai_c4_template.c [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+// Set up SPI slave select pin
+void keyboard_post_init_kb(void) {
+    #ifdef EXTERNAL_FLASH_SPI_SLAVE_SELECT_PIN
+	    setPinOutput(EXTERNAL_FLASH_SPI_SLAVE_SELECT_PIN);
+		writePinHigh(EXTERNAL_FLASH_SPI_SLAVE_SELECT_PIN);
+	#endif
+}
```
Comment (tzarc) [inline keyboards/custommk/bonsai_c4_template/bonsai_c4_template.c:23]: If this is a template for use in other boards, lets clean up the indenting and the like.

**Review by tzarc — CHANGES_REQUESTED** [review]
Needs conflicts fixed, too.

**Comment by zvecr** [comment]
Sorry, got conflicts again. Will merge asap once they are resolved.

**Review by zvecr — APPROVED** [review]

**Review by keyboard-magpie — APPROVED** [review]

**Pull request merged**